### DISCUSSION TITLE
The TinyBase Roadmap

### DISCUSSION THREAD

**Discussion by jamesgpearce** [tinyplex/tinybase#98] [Announcements]
Hey everyone! Whilst aspirational ideas do not always constitute hard commitments, I thought it would be valuable for you all to share the new [TinyBase roadmap](https://github.com/orgs/tinyplex/projects/1). Our community is growing and I owe it to you!

As you can see, the recent v4.x releases have been focussed on integrations (SQLite, Yjs, Automerge, PartyKit etc), and on the new React DOM components. I think there are a few more integrations to get through, including Postgres.

There has been a lot of interest in pushing the boundaries of TinyBase's opinionated data model. Today, a table cell can be a string, a number, or a boolean. But to model more interesting things (like many-to-many relationships) we may need to consider arrays, shallow objects, or nulls. This will take some design thought and will probably be breaking - so I feel like that is a good v5.0 theme.

If there are other things you would like to see TinyBase do in the future, file a task so I can add it to the board accordingly, and feel free to use this thread to discuss overall direction.

**Comment by WonderPanda** [comment]
I'm very excited about the proposed theme for 5.x! This could be a pretty big game changer for our current codebase where we have to do a lot of gymnastics to handle serializing objects and arrays as strings

**Comment by MentalGear** [comment]
I recently discovered tinybase and it's quite exciting, also for the CRDT roadmap. The team seems to have done tremendous work.

Yet I have some questions:

>  we may need to consider arrays, shallow objects, or nulls.

- Is it not currently possible to use arrays in tinybase?

- Can you add a migrations page to the guides?

- When is the ETA, are there docs available for v.5 ?

- CRDT: Data validation / schema
I meant to use vanilla yjs, but what worries me is that it doesn't have any schema or data validation, e.g. another peer could just overload a shared object with garbage data. There needs to be some way to limit size and type of the data stored, best with flexible validation rules like zod provides.
I guess here tinybase can help with the schema. But for the data validation part, I assume, if yjs was to be used with tinybase, one would still need to on the server manually decode and apply a peer's update to the whole data, validate the data, and accept or reject it. This is of course quite a lot of compute time for every update that is send from a client.

Since you seem to develop a own CRDT tech for v.5, I was wondering how you attempt to handle both those issues, which seem to be a big unsolved want for any CRDT yet.

**Comment by tnspacetime** [comment]
> Today, a table cell can be a string, a number, or a boolean. But to model more interesting things (like many-to-many relationships) we may need to consider arrays, shallow objects, or nulls. 

This would be amazing!! 

I think another major thing I would like to see is reactive queries!

**Reply by jamesgpearce** [reply]
I think queries should be reactive already! Feel free to share some code so I can see why they are not :)

**Comment by tnspacetime** [comment]
I think they are indeed reactive, for example by using

`useResultRowIds`

But I've noticed that when defining a query, for example as in 

const queries = [useCreateQueries](https://tinybase.org/api/ui-react/functions/queries-hooks/usecreatequeries/)(store, (store) => {
    console.log('Queries created');
    return [createQueries](https://tinybase.org/api/queries/functions/creation/createqueries/)(store).[setQueryDefinition](https://tinybase.org/api/queries/interfaces/queries/queries/methods/configuration/setquerydefinition/)(
      'dogColors',
      'pets',
      ({select, where}) => {
        select('color');
        where('species', 'dog');
      },
    );
  });
  
  and if I use `where(getCell => {})` with a complicated logic inside, then only `color` and `species` cells would be reactive, but not the cells used in the `where(getCell => {})` (or, for that matter, anything that lead to a change in the result of the queries unrelated to `color` and `species` won't trigger re-render).
